constraints/binRightmost.js

```diff
@@ -0,0 +1,60 @@
+/* Assign a violation for every node of category cat 
+such that its rightmost child of category (cat-1) 
+has more than two children.
+*/
+
+function binMaxRightmostBranches(s, ptree, cat) {
+  var vcount = 0;
+  //base case: we are at leaf && there are no children
+  //make sure there is children
+  if (ptree.children && ptree.children.length) {
+    if (ptree.cat === cat) {
+      //check rightmost child
+      var rightMost = ptree.children.length - 1;
+      var rightMostChild = ptree.children[rightMost];
+      if (rightMostChild.children && rightMostChild.children.length > 2) {
+        vcount++;
+      }       
+    }
+    //check other nodes in ptree
+    for(var i = 0; i < ptree.children.length; i++) {
+      vcount += binMaxRightmostBranches(s, ptree.children[i], cat);
+    }       
+  }
+  return vcount;
+};
+var parent_ptree;
+function binMaxRightmostLeaves(s, ptree, cat) {
+  //first find the first level of with cat --> recursive look into the children
+  var vcount = 0;
+  if(ptree.children && ptree.children.length) {
+  if(ptree.cat === cat) {
+    var recur_flag = 0;
+    for(var i = 0; i < ptree.children.length; i++) {
+      if(ptree.children[i].cat === cat) {
+        vcount++;
+	recur_flag = 1;
+	break;
+      }
+    }
+    if(recur_flag === 0) {
+      if(ptree.children.length > 2) {
+      	if(parent_ptree && ptree === parent_ptree.children[parent_ptree.children.length-1]) {
```

Comment:
There is no guarantee parent_ptree is correctly defined. One issue is that for the root node, parent_ptree will either be undefined or incorrectly defined (left over from a previous call to this function). The root node should automatically be considered rightmost. For example, according to the constraint definition, binMaxRightmostLeaves({}, {(a)(b)(c)}, "i") should return 1, since the iota dominates 3 phis. But instead it returns 0.